### PR TITLE
add hidefix

### PR DESCRIPTION
**Pull request by gauteh**
Checklist
- [X] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [X] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by gauteh** [comment]
@conda-forge-admin please rerender

**Comment by gauteh** [comment]
@conda-forge-admin, please ping conda-forge/staged-recipes PR ready for review!

**Comment by gauteh** [comment]
@conda-forge-admin, please ping conda-forge/staged-recipes do you have the chance to take a look?

File: conda-forge.yml [hunk 1]
```diff
@@ -1,5 +1,10 @@
-channel_priority: strict
-azure:
-  settings_win:
-    pool:
-      vmImage: windows-2019
\ No newline at end of file
+github:
+  branch_name: main
```
Comment (xhochy) [inline conda-forge.yml:2]: Please don't touch files outside of `recipes/`

File: recipes/hidefix/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    - {{ compiler('rust') }}
+  host:
+    - python
+    - pip
+    - maturin
+    - pkg-config
+  run:
+    - python
+    - libgcc-ng  # [ linux ]
+    - libcxx     # [ osx ]
```
Comment (xhochy) [inline recipes/hidefix/meta.yaml:35]: This should not be needed.
```suggestion
```

File: recipes/hidefix/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  build:
+    - cmake
+    - make
+    - pkg-config
+    - python
+    - cross-python_{{ target_platform }}  # [build_platform != target_platform]
+    - maturin                             # [build_platform != target_platform]
+    - {{ compiler('c') }}             # [ linux ]
+    - {{ compiler('m2w64_c') }}       # [ win ]
+    - {{ compiler('cxx') }}           # [ linux ]
```
Comment (xhochy) [inline recipes/hidefix/meta.yaml:25]:
```suggestion
    - {{ compiler('c') }}             # [linux]
    - {{ compiler('m2w64_c') }}       # [win]
    - {{ compiler('cxx') }}           # [linux]
```

File: recipes/hidefix/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+build:
+  number: 0
+
+requirements:
+  build:
+    - cmake
+    - make
+    - pkg-config
+    - python
```
Comment (xhochy) [inline recipes/hidefix/meta.yaml:20]:
```suggestion
    - python # [build_platform != target_platform]
```

**Review by xhochy — CHANGES_REQUESTED** [review]

File: conda-forge.yml (hunk 1 above)
Comment (gauteh) [reply conda-forge.yml:2]: I needed these changes for `roaring-landmask` to get it to build for arm64 and aarch64, but I can fix that later if it doesn't work.

File: recipes/hidefix/meta.yaml (hunk 2 above)
Comment (gauteh) [reply recipes/hidefix/meta.yaml:35]: I need those because of missing references to c++ stdlib. But can also try without.

**Comment by gauteh** [comment]
@conda-forge-admin please rerender

File: conda-forge.yml (hunk 1 above)
Comment (gauteh) [reply conda-forge.yml]: Removed these changes, think this should be ready for another pass. Thanks for your comments so far.

**Comment by gauteh** [comment]
Hi, @xhochy, @conda-forge/staged-recipes this is ready for another look.

**Comment by kmuehlbauer** [comment]
@conda-forge-admin, please rerender

**Comment by kmuehlbauer** [comment]
@conda-forge-admin, please restart ci

**Comment by gauteh** [comment]
Looks like all CI are passing.

File: recipes/hidefix/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    - {{ compiler('c') }}                 # [linux]
+    - {{ compiler('m2w64_c') }}           # [win]
+    - {{ compiler('cxx') }}               # [linux]
+    - {{ compiler('rust') }}
+  host:
+    - python
+    - pip
+    - maturin
+    - pkg-config
+  run:
```
[the remark this answers is not in the document]
Comment (gauteh) [reply recipes/hidefix/meta.yaml:32]: Thanks, fixed.

**Comment by kmuehlbauer** [comment]
@conda-forge/staged-recipes Please consider another review. To me it looks like it is ready to go. Thanks!

**Comment by ocefpaf** [comment]
@xhochy you were the first reviewer here. Have your comments being addressed?

File: recipes/hidefix/build.sh [hunk 6]
```diff
@@ -0,0 +1,19 @@
+# Based on the build.sh for the pysyntect-feedstock recipe.
+# https://github.com/conda-forge/pysyntect-feedstock/
+set -ex
+
+# Set conda CC as custom CC in Rust
+export CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_LINKER=$CC
+export CARGO_TARGET_X86_64_APPLE_DARWIN_LINKER=$CC
+export CARGO_TARGET_AARCH64_APPLE_DARWIN_LINKER=$CC
```
Comment (xhochy) [inline recipes/hidefix/build.sh:8]: This should already be covered by compiler('rust')

File: recipes/hidefix/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    - cmake
+    - make
+    - pkg-config
+    - python                              # [build_platform != target_platform]
+    - cross-python_{{ target_platform }}  # [build_platform != target_platform]
+    - maturin                             # [build_platform != target_platform]
+    - {{ compiler('c') }}                 # [linux]
+    - {{ compiler('m2w64_c') }}           # [win]
+    - {{ compiler('cxx') }}               # [linux]
+    - {{ compiler('rust') }}
```
Comment (weiji14) [inline recipes/hidefix/meta.yaml:26]: Need to add cargo-bundle-licenses too, see https://conda-forge.org/docs/maintainer/adding_pkgs.html#licenses-of-included-dependencies
```suggestion
    - {{ compiler('rust') }}
    - cargo-bundle-licenses
```

File: recipes/hidefix/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/hidefix-{{ version }}.tar.gz
+  sha256: 9ac6c81783e1d8c401c4c5bd2d2df1f6f85bc91de32889d54acf62506bbd6910
+
+build:
+  number: 0
```
Comment (weiji14) [inline recipes/hidefix/meta.yaml:13]: Will need to run `cargo-bundle-licenses` also. Do you prefer to add it in `recipe/meta.yaml` or in the `build.sh`/`bld.bat` script?

Option 1 is to do this, similar to https://github.com/conda-forge/lazrs-python-feedstock/blob/9954d7461d4a054855bd3f351d7d3d83cf4b2eb4/recipe/meta.yaml#L12-L15, but that means you'll lose the extra configuration in bld.bat/build.sh:

```suggestion
build:
  number: 0
  script:
    - cargo-bundle-licenses --format yaml --output THIRDPARTY.yml
    - {{ PYTHON }} -m pip install . -vv
```

Option 2 is to add it to `build.sh` and `bld.bat`, similar to https://github.com/conda-forge/polars-feedstock/blob/287e41db502984c1fe6528dded5aa1f57bea20d4/recipe/build.sh#L14.

File: recipes/hidefix/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  commands:
+    - pip check
+  requires:
+    - pip
+
+about:
+  home: https://github.com/gauteh/hidefix
+  summary: Concurrent HDF5 and NetCDF4 reader (experimental)
+  license: MIT
+  license_file: LICENSE
```
Comment (weiji14) [inline recipes/hidefix/meta.yaml:51]:
```suggestion
  license_file:
    - THIRDPARTY.yml
    - LICENSE
```

File: recipes/hidefix/conda_build_config.yaml [hunk 10]
```diff
@@ -0,0 +1,2 @@
+channel_sources:
+  - conda-forge/label/rust_dev,conda-forge
```
Comment (weiji14) [inline recipes/hidefix/conda_build_config.yaml:2]: Any reason in particular to get the rust_dev version?

File: recipes/hidefix/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    - cross-python_{{ target_platform }}  # [build_platform != target_platform]
+    - maturin                             # [build_platform != target_platform]
+    - {{ compiler('c') }}                 # [linux]
+    - {{ compiler('m2w64_c') }}           # [win]
+    - {{ compiler('cxx') }}               # [linux]
+    - {{ compiler('rust') }}
+  host:
+    - python
+    - pip
+    - maturin
```
Comment (weiji14) [inline recipes/hidefix/meta.yaml:30]: Looking at https://github.com/gauteh/hidefix/blob/v0.6.1/pyproject.toml#L2, you have maturin pinned to 0.14, so should probably do the same here:

```suggestion
    - maturin >=0.14,<0.15
```

**Review by weiji14 — COMMENTED** [review]
Hi @gauteh, just stumbled on this from the xarray backend issue, and would like to help nudge this forward a bit :D

File: recipes/hidefix/conda_build_config.yaml (hunk 10 above)
Comment (gauteh) [reply recipes/hidefix/conda_build_config.yaml:2]: I think hidefix requires nightly

File: recipes/hidefix/meta.yaml (hunk 8 above)
Comment (gauteh) [reply recipes/hidefix/meta.yaml:13]: If the build/bld files are not necessary that would be the best! I don't think I managed to get it to work without them, but maybe that has changed now?

**Comment by gauteh** [comment]
> Hi @gauteh, just stumbled on this from the xarray backend issue, and would like to help nudge this forward a bit :D

Thanks, much appreciated! The build stopped working while it was sitting in review during the spring, and I haven't had the chance to look into it during the summer. Any further ideas on how to fix it are much appreciated.

File: recipes/hidefix/meta.yaml (hunk 8 above)
Comment (weiji14) [reply recipes/hidefix/meta.yaml:13]: You could try to delete the build.sh/bld.bat files and commit the above change as a test? Can always revert back if needed.

File: recipes/hidefix/conda_build_config.yaml (hunk 10 above)
Comment (weiji14) [reply recipes/hidefix/conda_build_config.yaml:2]: Looks like rust_dev is currently at `1.72.0.dev20230606`, and main is at `1.71.1`. If you're after rust 1.72, then should be ok.

Comment (weiji14) [reply recipes/hidefix/conda_build_config.yaml:2]: Ok yeah, just tested compiling locally, and got this error when using the stable `rust`:
```python-traceback
  error: could not compile `hidefix` (lib) due to 8 previous errors
  💥 maturin failed
    Caused by: Failed to build a native library through cargo
    Caused by: Cargo build finished with "exit status: 101": `"cargo" "rustc" "--release" "--features" "static" "--features" "python" "--features" "extension-module" "--features" "fast-index" "--target" "x86_64-unknown-linux-gnu" "--manifest-path" "/home/conda/staged-recipes/build_artifacts/hidefix_1692063824569/work/Cargo.toml" "--message-format" "json" "--lib" "--crate-type" "cdylib"`
  error[E0554]: `#![feature]` may not be used on the stable release channel
    --> src/lib.rs:50:12
     |
  50 | #![feature(test)]
     |            ^^^^


  error[E0554]: `#![feature]` may not be used on the stable release channel
    --> src/lib.rs:51:12
     |
  51 | #![feature(cow_is_borrowed, array_methods)]
     |            ^^^^^^^^^^^^^^^


  error[E0554]: `#![feature]` may not be used on the stable release channel
    --> src/lib.rs:51:29
     |
  51 | #![feature(cow_is_borrowed, array_methods)]
     |                             ^^^^^^^^^^^^^


  error[E0554]: `#![feature]` may not be used on the stable release channel
    --> src/lib.rs:52:12
     |
  52 | #![feature(assert_matches)]
     |            ^^^^^^^^^^^^^^


  error[E0554]: `#![feature]` may not be used on the stable release channel
    --> src/lib.rs:53:12
     |
  53 | #![feature(slice_group_by)]
     |            ^^^^^^^^^^^^^^


  error[E0554]: `#![feature]` may not be used on the stable release channel
    --> src/lib.rs:54:12
     |
  54 | #![feature(mutex_unlock)]
     |            ^^^^^^^^^^^^


  error[E0554]: `#![feature]` may not be used on the stable release channel
    --> src/lib.rs:55:12
     |
  55 | #![feature(new_uninit)]
     |            ^^^^^^^^^^


  error: aborting due to 7 previous errors


  For more information about this error, try `rustc --explain E0554`.

  Error: command ['maturin', 'pep517', 'build-wheel', '-i', '/home/conda/staged-recipes/build_artifacts/hidefix_1692063824569/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/bin/python', '--compatibility', 'off'] returned non-zero exit status 1
  error: subprocess-exited-with-error
  
  × Building wheel for hidefix (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /home/conda/staged-recipes/build_artifacts/hidefix_1692063824569/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/bin/python /home/conda/staged-recipes/build_artifacts/hidefix_1692063824569/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py build_wheel /tmp/tmpoer5g6nf
  cwd: /home/conda/staged-recipes/build_artifacts/hidefix_1692063824569/work
  Building wheel for hidefix (pyproject.toml): finished with status 'error'
  ERROR: Failed building wheel for hidefix
Failed to build hidefix
ERROR: Could not build wheels for hidefix, which is required to install pyproject.toml-based projects
```
So can confirm that `rust_dev` is required.

File: recipes/hidefix/meta.yaml (hunk 8 above)
Comment (weiji14) [reply recipes/hidefix/meta.yaml:13]: > You could try to delete the build.sh/bld.bat files and commit the above change as a test? Can always revert back if needed.

I just tried compiling locally on Linux without the `build.sh` file, just the script commands above in `recipe/meta.yaml`, and it seemed to work. Not sure about macOS or Windows though.

Comment (gauteh) [reply recipes/hidefix/meta.yaml:13]: Trying it now. The feature configuration stuff has been moved into pyproject.toml.

File: recipes/hidefix/conda_build_config.yaml (hunk 10 above)
Comment (gauteh) [reply recipes/hidefix/conda_build_config.yaml:2]: Thanks for checking :+1:

File: recipes/hidefix/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+requirements:
+  build:
+    - cmake
+    - make
+    - pkg-config
+    - python                              # [build_platform != target_platform]
+    - cross-python_{{ target_platform }}  # [build_platform != target_platform]
+    - maturin                             # [build_platform != target_platform]
+    - {{ compiler('c') }}                 # [linux]
+    - {{ compiler('m2w64_c') }}           # [win]
```
Comment (weiji14) [inline recipes/hidefix/meta.yaml:27]: Hmm, looks like Windows builds are failing at https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=762255&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=2fb5b0a7-737c-5f5c-8f3f-f6db6174bacf&l=1746

```python-traceback
  error: failed to run custom build command for `pyo3-build-config v0.17.3`

  Caused by:
    process didn't exit successfully: `C:\bld\hidefix_1692082449997\work\target\release\build\pyo3-build-config-9474a066d1550e39\build-script-build` (exit code: 1)
    --- stdout
    cargo:rerun-if-env-changed=PYO3_CONFIG_FILE
    cargo:rerun-if-env-changed=PYO3_NO_PYTHON
    cargo:rerun-if-env-changed=PYO3_PYTHON
    cargo:rerun-if-env-changed=VIRTUAL_ENV
    cargo:rerun-if-env-changed=CONDA_PREFIX

    --- stderr
    error: failed to run the Python interpreter at C:\bld\hidefix_1692082449997\_build_env\python.exe: The system cannot find the file specified. (os error 2)
  warning: build failed, waiting for other jobs to finish...
  💥 maturin failed
    Caused by: Failed to build a native library through cargo
    Caused by: Cargo build finished with "exit code: 101": `"cargo" "rustc" "--release" "--features" "static" "--features" "python" "--features" "extension-module" "--features" "fast-index" "--manifest-path" "C:\\bld\\hidefix_1692082449997\\work\\Cargo.toml" "--message-format" "json" "--lib" "--crate-type" "cdylib"`
  Error: command ['maturin', 'pep517', 'build-wheel', '-i', 'C:\\bld\\hidefix_1692082449997\\_h_env\\python.exe', '--compatibility', 'off'] returned non-zero exit status 1
  error: subprocess-exited-with-error
  
  Building wheel for hidefix (pyproject.toml) did not run successfully.
  exit code: 1
  
  See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: 'C:\bld\hidefix_1692082449997\_h_env\python.exe' 'C:\bld\hidefix_1692082449997\_h_env\lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py' build_wheel 'C:\Users\VSSADM~1\AppData\Local\Temp\tmpuhibrxi7'
  cwd: C:\bld\hidefix_1692082449997\work
  Building wheel for hidefix (pyproject.toml): finished with status 'error'
  ERROR: Failed building wheel for hidefix
Failed to build hidefix
ERROR: Could not build wheels for hidefix, which is required to install pyproject.toml-based projects
Exception information:
Traceback (most recent call last):
  File "%PREFIX%\lib\site-packages\pip\_internal\cli\base_command.py", line 180, in exc_logging_wrapper
    status = run_func(*args)
  File "%PREFIX%\lib\site-packages\pip\_internal\cli\req_command.py", line 248, in wrapper
    return func(self, options, args)
  File "%PREFIX%\lib\site-packages\pip\_internal\commands\install.py", line 429, in run
    raise InstallationError(
pip._internal.exceptions.InstallationError: Could not build wheels for hidefix, which is required to install pyproject.toml-based projects
Removed build tracker: 'C:\\Users\\VssAdministrator\\AppData\\Local\\Temp\\pip-build-tracker-rq1v8b8t'
Traceback (most recent call last):
  File "D:\a\1\s\.ci_support\build_all.py", line 218, in <module>
    build_all(os.path.join(root_dir, "recipes"), args.arch)
  File "D:\a\1\s\.ci_support\build_all.py", line 118, in build_all
    build_folders(recipes_dir, folders, arch, channel_urls)
  File "D:\a\1\s\.ci_support\build_all.py", line 174, in build_folders
    conda_build.api.build([recipe], config=get_config(arch, channel_urls))
  File "C:\Miniforge\lib\site-packages\conda_build\api.py", line 253, in build
    return build_tree(
  File "C:\Miniforge\lib\site-packages\conda_build\build.py", line 3799, in build_tree
    packages_from_this = build(
  File "C:\Miniforge\lib\site-packages\conda_build\build.py", line 2624, in build
    windows.build(
  File "C:\Miniforge\lib\site-packages\conda_build\windows.py", line 341, in build
    check_call_env(
  File "C:\Miniforge\lib\site-packages\conda_build\utils.py", line 450, in check_call_env
    return _func_defaulting_env_to_os_environ("call", *popenargs, **kwargs)
  File "C:\Miniforge\lib\site-packages\conda_build\utils.py", line 426, in _func_defaulting_env_to_os_environ
    raise subprocess.CalledProcessError(proc.returncode, _args)
subprocess.CalledProcessError: Command '['cmd.exe', '/d', '/c', 'conda_build.bat']' returned non-zero exit status 1.
##[error]Cmd.exe exited with code '1'.
##[section]Finishing: Build recipe
```

Seems to be failing after you removed the `export CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_LINKER=$CC` lines in commit daed7af03c86414f8575a1dc7091fd1e707bf2eb. But you've already set `- {{ compiler('rust') }}`, so I'm wondering if this `{{ compiler('m2w64_c') }}` line is the issue?

Comment (gauteh) [reply recipes/hidefix/meta.yaml:27]: Trying now. By the way, if you have access to push to this branch feel free to go ahead. I'm working on getting v0.6.4 of hidefix out with updated dependencies. That tends to make building easier as well.

Comment (weiji14) [reply recipes/hidefix/meta.yaml:27]: Nope, I'm not a conda-forge core maintainer, so don't have push permissions. But it looks like you're on the right track!

File: recipes/hidefix/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    - python
+    - pip
+    - maturin >=0.15,<0.16
+    - pkg-config
+  run:
+    - python
+    - xz
+    - numpy
+    - xarray
+    - netCDF4
```
Comment (ocefpaf) [inline recipes/hidefix/meta.yaml:41]: Not sure if this will fix Windows but it is worth a try:

```suggestion
    - netCDF4
    - ucrt  # [win]
```

Comment (gauteh) [reply recipes/hidefix/meta.yaml:41]: I think it is missing the windows cxx runtime.

File: recipes/hidefix/meta.yaml (hunk 12 above)
Comment (gauteh) [reply recipes/hidefix/meta.yaml:27]: Maybe this is related to the `ucrt` runtime, maybe we can just use compiler('cxx') on win as well.

File: recipes/hidefix/meta.yaml [hunk 14]
```diff
@@ -39,7 +39,8 @@ requirements:
     - numpy
     - xarray
     - netCDF4
-    - ucrt  # [win]
+    - ucrt          # [win]
+    - vc14_runtime  # [win]
```
Comment (ocefpaf) [inline recipes/hidefix/meta.yaml:43]: I'd try one and then the other. Both at the same time you may never know which one is really needed and/or why it is still broken.

Comment (gauteh) [reply recipes/hidefix/meta.yaml]: Yes. Only ucrt fixed the ucrt error. I'm trying with the compiler cxx template on win as well.

**Comment by gauteh** [comment]
@conda-forge-admin please rerender

**Review by weiji14 — APPROVED** [review]
Yay, Windows tests are passing again :tada: @xhochy and @conda-forge/staged-recipes, this looks ready for final review and merge. Thanks @gauteh for working through the build errors!

**Comment by gauteh** [comment]
> Yay, Windows tests are passing again 🎉 @xhochy and @conda-forge/staged-recipes, this looks ready for final review and merge. Thanks @gauteh for working through the build errors!

Thanks. The package now runs all pytest tests, so I'm pretty confident it is working as it should.

**Comment by gauteh** [comment]
Hi, any chance of pushing this through soon? Things tend to change and need to be maintained anyway, so if it is sitting too long it will require new changes and a new review.

Thanks for your time, Gaute

**Comment by gauteh** [comment]
Ping, please take a look when you have time.

Regards, Gaute

**Pull request closed without merging**